Migrate tracer simplex example (#365)
lethe-cfd/lethe: doc/source/examples/multiphysics/tracer-through-cad-junction/tracer-through-cad-junction.rst

======================================
Tracer through CAD Junction in Simplex
======================================

In this example, we solve a problem using the simplex capabilities of Lethe. The simplex grids can be generated easily to represent complex geometries, such as a junction of channels made a boolean addition. The tracer physics usage will also be demonstrated.

Features
----------------------------------

> - Solvers: ``gls_navier_stokes_3d`` 

Author: ```suggestion
- Solver: ``gls_navier_stokes_3d`` 
```

.. image:: images/salome_mesh_creation.png
    :alt: Contextual menu for mesh generation in Salome
    :align: center

Adding the boundary IDs from Gmsh
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

A criterion for the use of ``.msh`` file in Lethe is that the boundaries have an associated ID number. These numbers can be added in Gmsh.

The boundaries from the mesh must be linked to a Physical object. In 3D for example, the inlet, outlet and walls must be specified as Physical surfaces. Selecting the surface for the inlet and outlet can be done easily, and then the ``.geo`` file can be text-edited to add a new physical surface with the remaining surfaces. The physical volume must also be specified.
In this specific case, there are four boundaries: inlets with no tracer (1), inlet with tracer (2), walls (3) and outlets (4).

.. code-block:: text

    Merge "Mesh_1.mesh";
    Physical Surface(1) = {729, 712};
    Physical Surface(2) = {10};
    Physical Surface(3) = {1:9,11:711,713:728,730:848,850,852:5000};
    Physical Surface(4) = {849, 851};

>     Physical Volume(1) = {1};

Reviewer: ```suggestion

The ``Physical Volume`` must also be specified for compatibility with deal.II, as stated with this line from the ``.geo`` file:

.. code-block:: text

    Physical Volume(1) = {1};

```

I think it will be clearer this way.

Author: Good suggestion!

The last step in Gmsh is to generate the 3D mesh, and then save it to a ``.msh`` file.

Parameter file
--------------

Using Lethe requires a solver executable, in this case ``gls_navier_stokes_3d``, and a ``.prm`` file. This second one can be setup in many ways, but for this specific case two aspects must be treated with more care: the enabling of simplex mode, and the setup of the tracer physics. 

Enabling the Simplex mode
~~~~~~~~~~~~~~~~~~~~~~~~~

> The simplex mode can be used by setting a boolean to true and making sure the ``.msh`` file used has the proper type of elements. 

Reviewer: A bit vague, maybe:
```suggestion
In the ``subsection mesh``, the simplex mode can be enabled with ``set simplex = true``. Also, the mesh in the ``.msh`` file must obviously be built with simplex elements.
```

.. code-block:: text

    #---------------------------------------------------
    # Mesh
    #---------------------------------------------------
    subsection mesh
        set type                 = gmsh
        set file name            = vessels.msh
        set initial refinement   = 0
        set simplex              = true
    end

.. warning:: 

>     If the mesh type is set to ``dealii`` instead of ``gmsh``, making sure the imported mesh is built with simplices doesn't make sense; the deal.II generated mesh will be built properly.

Reviewer: ```suggestion
    If the mesh type is set to ``dealii`` instead of ``gmsh``, the deal.II generated mesh will be properly built with simplices.
```

.. warning:: 

>     It is crucial to consider a current (2022-01-19) limitation : simplex grid refinement isn't implemented in Lethe.

Reviewer: ```suggestion
    [2022-01-19] It is crucial to consider a current limitation : simplex grid refinement isn't implemented in Lethe.
```

    # Physical Properties
    #---------------------------------------------------
    subsection physical properties
        set kinematic viscosity            = 1
        set tracer diffusivity             = 0.001
    end

.. code-block:: text

    #---------------------------------------------------
    # FEM
    #---------------------------------------------------
    subsection FEM
        set velocity order            = 1
        set pressure order            = 1
        set tracer order              = 1
    end

.. code-block:: text

    # --------------------------------------------------
    # Boundary Conditions
    #---------------------------------------------------
    subsection boundary conditions tracer
    set number                  = 2
        subsection bc 0
        set id = 1
        set type              = dirichlet
            subsection dirichlet
                    set Function expression =  0 
            end
        end
        subsection bc 1
        set id = 2
        set type              = dirichlet
            subsection dirichlet
                    set Function expression =  1 
            end
        end
    end

> The boundary conditions are written in a specific way. We have a number of specified 2 boundaries, a Dirichlet condition of 1 for an inlet, and another Dirichlet condition of 0 for the two other inlets; all the remaining boundaries are unspecified. An unspecified boundary condition in Lethe is considered as the natural condition of finite elements, which is a zero Neumann condition. This is the condition needed in that case for the walls and outlets: the walls must not let any tracer out, and the outlets must not apply any constraint on their tracer level.

Reviewer 2: adding the difference between the normal boundary condition subsection and the tracer boundary conditions subsection would be helpful I think

Results
----------

> The case must be run with the solver and the parameter file. Then, the results can be viewed using software such as Paraview.

Reviewer: add the run command, if on the example folder? (would not hurt I think)

.. image:: images/paraview_tracer.png
    :alt: Simulation results in Meshgrid format

>     :align: center

Reviewer: could you add an interpretation of these results? What are we looking at? What does it mean physically?

.. note:: 
    The ``boundary conditions tracer`` subsection is different from the general ``boundary conditions`` 
    which concerns the flow. 

This ``boundary conditions`` subsection for the flow is setup as follows. The inlet with a high tracer concentration (``id = 2``)
is given a higher velocity than the other two (``id = 1``). The walls of the junction (``id = 3``) are given a ``no slip`` type.
The remaining boundaries (``id = 4``) are unspecified for the same reason as in the previous subsection: no constraint 
must be applied to the outlet flow. 

.. code-block:: text

    # --------------------------------------------------
    # Boundary Conditions
    #---------------------------------------------------
    subsection boundary conditions
        set number                  = 3
        subsection bc 0
            set id = 1
            set type              = function
            subsection u
                set Function expression = 0
            end
            subsection v
                set Function expression = 0
            end
            subsection w
                set Function expression = 1
            end
        end

>         # boundary id2 will have the tracer

Author: ```suggestion
        # The boundary id = 2 will have the tracer
```
